include/deal.II/matrix_free/fe_evaluation.h

```diff
@@ -6413,39 +6432,87 @@ FEEvaluationBase<dim, n_components_, Number, is_face, VectorizedArrayType>::
           VectorizedArrayType tmp[dim][dim];
           for (unsigned int i = 0; i < dim; ++i)
             for (unsigned int j = 0; j < dim; ++j)
-              tmp[i][j] = 0.;
+              {
+                tmp[i][j] = hessian_in[comp][i][0] * jac[0][j];
+                for (unsigned int k = 1; k < dim; ++k)
+                  tmp[i][j] += hessian_in[comp][i][k] * jac[k][j];
+              }
 
+          // 2. hessian_unit = J^T * tmp
+          VectorizedArrayType tmp2[dim][dim];
           for (unsigned int i = 0; i < dim; ++i)
             for (unsigned int j = 0; j < dim; ++j)
-              for (unsigned int k = 0; k < dim; ++k)
-                tmp[i][j] += hessian_in[comp][i][k] * jac[k][j] * JxW;
-
-          // 2. hessian_unit = J^T * tmp
+              {
+                tmp2[i][j] = jac[0][i] * tmp[0][j];
+                for (unsigned int k = 1; k < dim; ++k)
+                  tmp2[i][j] += jac[k][i] * tmp[k][j];
+              }
 
           // diagonal part
           for (unsigned int d = 0; d < dim; ++d)
-            {
-              hessians_quad[(comp * hdim + d) * nqp + q_point] = 0;
-              for (unsigned int i = 0; i < dim; ++i)
-                hessians_quad[(comp * hdim + d) * nqp + q_point] +=
-                  jac[i][d] * tmp[i][d];
-            }
+            hessians_quad[(comp * hdim + d) * nqp + q_point] = tmp2[d][d] * JxW;
 
-          // off diagonal part (use symmetry)
-          for (unsigned int d = 1, off_dia = dim; d < dim; ++d)
-            for (unsigned int e = 0; e < d; ++e, ++off_dia)
-              {
-                hessians_quad[(comp * hdim + off_dia) * nqp + q_point] = 0;
-                for (unsigned int i = 0; i < dim; ++i)
-                  hessians_quad[(comp * hdim + off_dia) * nqp + q_point] +=
-                    2.0 * jac[i][e] * tmp[i][d];
-              }
+          // off diagonal part
+          for (unsigned int d = 0, off_diag = dim; d < dim; ++d)
+            for (unsigned int e = d + 1; e < dim; ++e, ++off_diag)
+              hessians_quad[(comp * hdim + off_diag) * nqp + q_point] =
+                (tmp2[d][e] + tmp2[e][d]) * JxW;
         }
     }
   else
     {
-      // TODO: implement general path submit_hessian
-      AssertThrow(false, ExcNotImplemented());
+      const Tensor<2, dim, VectorizedArrayType> jac = this->jacobian[q_point];
+      const VectorizedArrayType                 JxW = this->J_value[q_point];
```

Comment:
```suggestion
      const Tensor<2, dim, VectorizedArrayType>& jac = this->jacobian[q_point];
      const VectorizedArrayType&                 JxW = this->J_value[q_point];
```

Reply:
We are a bit inconsistent here, as we do https://github.com/dealii/dealii/blob/b6bd942d5fb748f34e3c5b101ad198083cb2d8bd/include/deal.II/matrix_free/fe_evaluation.h#L5805-L5808 vs https://github.com/dealii/dealii/blob/master/include/deal.II/matrix_free/fe_evaluation.h#L6279-L6282
In the end, it will not matter much because a sane compiler will likely treat both cases similarly and avoid temporaries as much as possible. I am not sure if aliasing could force the compiler to emit unnecessary loads in the case of a reference and if that was the reason to do a copy in the second appearance for `submit_gradient`. I suggest we leave this as is (no reference) and just do something in case we have verified the reference does not lead to unnecessary loads in the multi-component case in the assembly code.

Reply:
You could show me how to look for this in assembly code.